evalml/automl/automl_search.py

```diff
@@ -403,143 +404,154 @@ def search(self, data_checks="auto", show_iteration_plot=True):
+                logger.info(f"Searching up to {self.max_batches} batches for a total of {self.max_iterations} pipelines. ")
+            elif self.max_iterations is not None:
+                logger.info("Searching up to %s pipelines. " % self.max_iterations)
+            if self.max_time is not None:
+                logger.info("Will stop searching for new pipelines after %d seconds.\n" % self.max_time)
+            logger.info("Allowed model families: %s\n" % ", ".join([model.value for model in self.allowed_model_families]))
+            search_iteration_plot = None
+            if self.plot:
+                search_iteration_plot = self.plot.search_iteration_plot(interactive_plot=show_iteration_plot)
+
+            self._start = time.time()
+
+            should_terminate = self._add_baseline_pipelines()
+            if should_terminate:
+                return
 
-        elapsed_time = time_elapsed(self._start)
-        desc = f"\nSearch finished after {elapsed_time}"
-        desc = desc.ljust(self._MAX_NAME_LEN)
-        logger.info(desc)
+            current_batch_pipelines = []
+            current_batch_pipeline_scores = []
+            while self._check_stopping_condition(self._start):
+                try:
+                    if current_batch_pipeline_scores and np.isnan(np.array(current_batch_pipeline_scores, dtype=float)).all():
+                        raise AutoMLSearchException(f"All pipelines in the current AutoML batch produced a score of np.nan on the primary objective {self.objective}.")
+                    current_batch_pipelines = self._automl_algorithm.next_batch()
+                    current_batch_pipeline_scores = []
+                except StopIteration:
+                    logger.info('AutoML Algorithm out of recommendations, ending')
+                    break
+
+                current_batch_size = len(current_batch_pipelines)
+                current_batch_pipeline_scores = self._evaluate_pipelines(current_batch_pipelines, search_iteration_plot=search_iteration_plot)
+
+                # Different size indicates early stopping
+                if len(current_batch_pipeline_scores) != current_batch_size:
+                    break
+
+            elapsed_time = time_elapsed(self._start)
+            desc = f"\nSearch finished after {elapsed_time}"
+            desc = desc.ljust(self._MAX_NAME_LEN)
+            logger.info(desc)
+
+            best_pipeline = self.rankings.iloc[0]
+            best_pipeline_name = best_pipeline["pipeline_name"]
+            self._find_best_pipeline()
+            logger.info(f"Best pipeline: {best_pipeline_name}")
+            logger.info(f"Best pipeline {self.objective.name}: {best_pipeline['score']:3f}")
+            self._searched = True
+        else:
+            logger.info("AutoMLSearch has already been run and will not run again on the same instance. Re-initialize AutoMLSearch to search again.")
 
+    def _find_best_pipeline(self):
```

Comment:
New method to find the best pipeline and train only if the pipeline is new

Reply:
Nice